components/edit.js

```diff
@@ -0,0 +1,235 @@
+    if (isEdit && !edit) {
+      setEditing(true)
+    }
+
+    // If the URL has us not editing but we just were, set to false, and update props
+    if (!isEdit && edit) {
+      setEditing( false )
+      router.replace(router.asPath)
+    }
+  }, [edit, router])
+
+  const editpage = <EditPage item={item}/>
+  const editlink = <EditLink item={item}/>
+
+  const componentToShow = edit ? editpage :
+  <>
+    {viewComponent}
+    {editlink}
+  </>
+  return componentToShow
+}
+
+export const EditableItem = ({ item, children }) => {
+  return useEditor(item, children)
+}
+
+EditableItem.propTypes = {
+  item: PropTypes.object.isRequired,
+}
+
+const SaveCancelControl = ({
+  cancelClick,
+  saveClick,
+  itemPath,
+  saveEnabled,
+}) => {
+  return (
+    <div className="flex space-x-1">
+      <Button.Link href={itemPath} navigationClick={cancelClick}>
+        Cancel
+      </Button.Link>
+      <Button
+        onClick={saveClick}
+        enabled={saveEnabled}
+        type={saveEnabled ? "primary" : "info"}
+      >
+        Save
+      </Button>
+    </div>
+  )
+}
+
+SaveCancelControl.propTypes = {
+  cancelClick: PropTypes.func.isRequired,
+  saveClick: PropTypes.func.isRequired,
+  itemPath: PropTypes.string.isRequired,
+  saveEnabled: PropTypes.bool.isRequired,
+}
+
+function sortedJson(obj) {
+  if (Array.isArray(obj)) {
+    return obj.map((value) => sortedJson(value))
+  }
+  // We know it's not an array if we're here because the above `if`
+  if (typeof obj == "object") {
+    const sorted = {}
+    Object.keys(obj)
+      .sort()
+      .forEach((key) => {
+        sorted[key] = obj[key]
+      })
+    return sorted
+  }
+  return obj
+}
+
+const SavedErrors = ({ errors = [] }) => {
+  return (
+    <div>
+      <p className="text-lg text-rose-600">Errors from Save:</p>
+      <ul className="list-disc">
+        {errors.map((error) => (
+          <li key={error.description} className="text-base text-rose-600">
+            {error.description}
+          </li>
+        ))}
+      </ul>
+    </div>
+  )
+}
+
+SavedErrors.propTypes = {
+  errors: PropTypes.array.isRequired,
+}
+
+const EditPage = ({ item }) => {
+  const path = item["@id"]
+
+  const loggedIn = useAuthenticated()
+
+  const editable = (session, item) => {
+    // cannot edit if not logged in or object not editable
+    const editable = canEdit(item)
+    return loggedIn && editable
+  }
+
+  const jsonErrors = (json) => {
+    // cannot save if we cannot edit or if the JSON is wrong
+    try {
+      JSON.parse(json)
+      return []
+    } catch (err) {
+      // Save the error
+      return [err.message]
+    }
+  }
+
+  const { session } = useContext(SessionContext)
+
+  const [text, setText] = useState(() => {
+    JSON.stringify({}, null, 4)
+  })
+
+  const [saveErrors, setErrors] = useState([])
+
+  const [editorStatus, setEditorStatus] = useState({
+    canEdit: editable(session, item),
+    canSave: editable(session, item),
+    errors: [],
+  })
```

Comment:
By initializing a state with an object that calls functions, those functions get called at every render, but their results get thrown away after the first render.

Reply:
Another way to do this is to pass a function to `useState()` so that the callback calls a function that only gets executed the first render. But for your case, I think what you have here is better. But for other cases where a function initializes a state, you can do:
```
const [state, setState] = useState(() => f())
```
So in this example, `f()` only gets called on the first render, and its return value initializes `state`.